pandas/tools/merge.py

```diff
@@ -1361,15 +1361,18 @@ def __init__(self, objs, axis=0, join='outer', join_axes=None,
             objs = [obj for obj in objs if obj is not None]
         else:
             # #1649
-            clean_keys = []
+            clean_keys_list = []
             clean_objs = []
             for k, v in zip(keys, objs):
                 if v is None:
                     continue
-                clean_keys.append(k)
+                clean_keys_list.append(k)
                 clean_objs.append(v)
             objs = clean_objs
-            keys = clean_keys
+            clean_keys_index = Index(clean_keys_list)
+            if isinstance(keys, Index):
```

Comment:
Seems like you should be able to change the name of `clean_keys_list` back to `clean_keys` and

```python
if isinstance(keys, Index):
    keys = Index(clean_keys, name=keys.name)
```